x/qgb/orchestrator/parser.go

```diff
@@ -0,0 +1,96 @@
+package orchestrator
+
+import (
+	"fmt"
+
+	"github.com/celestiaorg/celestia-app/app"
+	qgbtypes "github.com/celestiaorg/celestia-app/x/qgb/types"
+	sdkcodec "github.com/cosmos/cosmos-sdk/codec"
+	sdkcodectypes "github.com/cosmos/cosmos-sdk/codec/types"
+	"github.com/cosmos/cosmos-sdk/std"
+	sdktypes "github.com/cosmos/cosmos-sdk/types"
+	sdktx "github.com/cosmos/cosmos-sdk/types/tx"
+	"github.com/gogo/protobuf/proto"
+	coretypes "github.com/tendermint/tendermint/types"
+)
+
+var _ QGBParserI = QGBParser{}
+
+type QGBParserI interface {
+	IsDataCommitmentConfirm(msg sdktypes.Msg) (bool, error)
+	ParseDataCommitmentConfirm(msg sdktypes.Msg) (qgbtypes.MsgDataCommitmentConfirm, error)
+	IsValsetConfirm(msg sdktypes.Msg) (bool, error)
+	ParseValsetConfirm(msg sdktypes.Msg) (qgbtypes.MsgValsetConfirm, error)
+	ParseCoreTx(tx coretypes.Tx) (sdktx.Tx, error)
+	ParseSdkTx(any *sdkcodectypes.Any) (sdktypes.Msg, error)
+}
+
+type QGBParser struct {
+	codec sdkcodec.Codec
+}
+
+func NewQGBParser(codec sdkcodec.Codec) *QGBParser {
+	return &QGBParser{codec: codec}
+}
+
+func (parser QGBParser) IsValsetConfirm(msg sdktypes.Msg) (bool, error) {
+	switch msg.(type) {
+	case *qgbtypes.MsgValsetConfirm:
+		return true, nil
+	default:
+		return false, nil
+	}
+}
+
+func (parser QGBParser) IsDataCommitmentConfirm(msg sdktypes.Msg) (bool, error) {
+	switch msg.(type) {
+	case *qgbtypes.MsgDataCommitmentConfirm:
+		return true, nil
+	default:
+		return false, nil
+	}
+}
+
+func (parser QGBParser) ParseValsetConfirm(msg sdktypes.Msg) (qgbtypes.MsgValsetConfirm, error) {
+	vs, ok := msg.(*qgbtypes.MsgValsetConfirm)
+	if !ok {
+		return qgbtypes.MsgValsetConfirm{},
+			fmt.Errorf("unable to cast sdk message to msg valset confirm")
+	}
+	return *vs, nil
+}
+
+func (parser QGBParser) ParseDataCommitmentConfirm(msg sdktypes.Msg) (qgbtypes.MsgDataCommitmentConfirm, error) {
+	dcc, ok := msg.(*qgbtypes.MsgDataCommitmentConfirm)
+	if !ok {
+		return qgbtypes.MsgDataCommitmentConfirm{},
+			fmt.Errorf("unable to cast sdk message to msg data commitment confirm")
+	}
+	return *dcc, nil
+}
+
+func (parser QGBParser) ParseCoreTx(tx coretypes.Tx) (sdktx.Tx, error) {
+	var sdkTx sdktx.Tx
+	err := proto.Unmarshal(tx, &sdkTx)
+	if err != nil {
+		return sdktx.Tx{}, fmt.Errorf("error while unmarshalling core transaction: %s", err)
+	}
+	return sdkTx, nil
+}
+
+func (parser QGBParser) ParseSdkTx(any *sdkcodectypes.Any) (sdktypes.Msg, error) {
```

Comment:
Messages and transactions are different. A transaction is the following:
https://github.com/celestiaorg/cosmos-sdk/blob/482d8986c7842b93f40239a210fac8b3975db39a/types/tx/tx.pb.go#L33

which contains a body:
https://github.com/celestiaorg/cosmos-sdk/blob/482d8986c7842b93f40239a210fac8b3975db39a/types/tx/tx.pb.go#L351

containing messages of type `Any`:
https://github.com/celestiaorg/cosmos-sdk/blob/482d8986c7842b93f40239a210fac8b3975db39a/types/tx/tx.pb.go#L359

Also, for more information, a cosmos-sdk transaction is different from a tendermint transaction. The latter is basically `[]byte` while the former contains a lot of fields.

The reason I am using the terminology  `ParseSdkTx` is to reflect that this will be related to parsing what's inside a cosmos-sdk sdk transaction, and not a tendermint one.

You are right, for the naming I will create an issue to update it later. If I make this change in here, and I built a lot on this method name, I will go through a nightmare of conflicts haha 

Issue: https://github.com/celestiaorg/celestia-app/issues/879

Reply:
fixed [93dffaa](https://github.com/celestiaorg/celestia-app/pull/878/commits/93dffaa3b4f36a1ed12369ae4041fc4246d74dd6)